Fix. Leave group actions

Celandriel opened the pull request:
Fix for issue #1768 where the bot master was not getting reset. 
I also cleaned up leave group action to focus up function scope.
I moved the resets from #612 to the actual leaving function.

Disclosure: LLMs were NOT used in authoring this PR.

Test cases to consider for testers.
1. Master disbands group with random bots. Bots should go about their business. 
2. If you can dual box, create a raid where two real player both invite random bots. One player leaves group, their bots should also leave. (edge case, if a random bot that is supposed to leave the group becomes leader they may disband the whole group.

Celandriel commented:
I thought the two real players could be a problem.
Some clarifying questions. 
Is this with only random bots or altbots as well?
Alt bots and addClass bots are intentionally excluded from leaving and must be logged out by command. 

Did the first leader invite all random bots? Or did the first add some and the second more? And then the seconds bots remain in group? 

Edit: Just saw the video. 
It looks like the master is being passed on and not reset correctly. Almost as if they do not all receive the same message. 
Will investigate triggers further.

Celandriel commented:
I'm going to stop trying to fix this from my phone.
This commit should be the one to consider if anyone feels like testing before I get to it.
https://github.com/mod-playerbots/mod-playerbots/pull/1774/commits/a9662f710f3d9ad4e9109aac4de7d9e153c8d95e

hermensbas commented on src/PlayerbotAI.cpp:
```diff
@@ -1363,16 +1362,15 @@ void PlayerbotAI::DoNextAction(bool min)
+    
+    PlayerbotAI* botAI = GET_PLAYERBOT_AI(bot);
+    if (!botAI)
+    {
```
Even though i personally prefer curly braces the code guideline states no curly brackets in this scenario.

Celandriel commented:
This isnt working correctly right now.

nl-saw commented on src/PlayerbotAI.cpp:
```diff
@@ -1363,67 +1365,15 @@ void PlayerbotAI::DoNextAction(bool min)
     if (master)
         masterBotAI = GET_PLAYERBOT_AI(master);
-
+    
```
Sneaky tab

nl-saw commented:
> I think this section
> 
> https://github.com/Celandriel/mod-playerbots/blob/aa6a7bfc997f7d654253e4cfc9ce81e8868df289/src/PlayerbotAI.cpp#L1369-L1400
> Should be merged with UpdateAIGroupMembership to have a single manager of groups and masters instead of competing actions trying to modify the master/group state

Yeah I think that would make sense.

Also given that in
```cpp
    // Test BG master set
    if (group && (!master || (masterBotAI && !masterBotAI->IsRealPlayer())))
```

this may have been some test at some point, didn't dive in the full history, but indeed having things compete in this manner is not great at all.

Celandriel commented:
Just pushed the changes. 
From the testing ive done it seems acceptable. 
I have not specifically tested the Alterac Valley case that the code considers, or the LFG, but from a simple invite/uninvite/leave perspective it seems appropriate.

One big change in the group logic was that previously the UpdateAIGroupMembership only iterated through all bots when in LFG not normal groups. This changes the logic to always call search for another real player. (Except in AV where other bots are acceptable masters) and relegates all logic to finding a new master to the new helper FindNewMaster. I did this so that I can check on triggering a group leave action and deciding if a bot should leave. 
This must be done in both places because you cannot predict which event happens first (processing the message, or update cycle for the bot), so you need redundancy to not leave a group prematurely. In the future this could be extended to logout non alts and addclass, but that is out of scope.

ScoobyPwnsOnU commented:
Tested in multiple scenarios.

1 player and 2 rndbots:
They followed me -> i left party -> they both stood there for about 5 seconds then took off in 2 different directions

1 player and 4 altbots:
They followed me -> i left party -> they kept following me

1 player and 4 addclassbots:
They followed me -> i left party -> they kept following me

1 player, 2 rndbots, 1 altbot, and 1 addclassbot:
They followed me -> i left party -> the altbot and addclassbot kept following me, the 2 rndbots wandered off

3 players multiboxing, 1 altbot, and 1 rndbot:
They followed me -> i left party, the altbot dropped party also -> the party is now 2 players and an rnd bot -> the rnd bot starts following the new party leader -> i leave party on that new leader -> the party is now 1 player and 1 rndbot -> the rnd bot starts following the last player in the party and newest party leader -> i leave party on that player, leaving the rndbot with nobody left -> the rnd bot wanders off

All of these scenarios sound like they played out correctly. ✅

Wishmaster117 commented:
@Celandriel , I'm not sure I fully understand the logic, but it seems to me there's some confusion in "Player* PlayerbotAI::FindNewMaster()".

GET_PLAYERBOT_AI(groupLeader) returns nullptr for a real (non-bot) player.
So if the leader is a real player, leaderBotAI == nullptr, the condition (!leaderBotAI || !leaderBotAI->IsRealPlayer()) is true, the leader isn't assigned as the new master, and we go looking for someone else.

Conversely, we only take the leader if leaderBotAI exists and leaderBotAI->IsRealPlayer() is true, which is strange (a "real" AI bot?).

As a result, the real leader is almost never chosen, which will cause erratic transitions (or a lack of reassignment) in two-player raids.

After that, it seems to me there's some confusion between `playerMaster` and `newMaster`. In the loop, we have:

* If `memberBotAI` exists (meaning `member` is a bot), we test `memberBotAI->IsRealPlayer()` to set `playerMaster`. This will never trigger (a bot is not `IsRealPlayer()`), so `playerMaster` remains useless/unassigned in these branches.

* For real players, where `memberBotAI == nullptr`, we directly assign `newMaster = member; break;` (outside of Battlegrounds), without prioritizing the leader or checking the lead transfer. We then end up taking the first real member found, which doesn't seem to be the desired policy.

The AV (Battlegrounds) case is too nested; okay for now, but the leader priority should remain clear.
The AV block is specific and can remain, but the priority rule outside of Battlegrounds should be:
"If the leader is a valid real player, we take the leader otherwise, no opportunistic reassignment." .

Am I wrong? I haven't tested the modifications, just read the code.

nl-saw commented:
Overall, the function works for most cases, but there are a few areas that might benefit from some refactoring and clearer intent. None of these are blockers, just things to consider for maintainability and readability (with summary of LLM):

Null Safety:
Some paths (e.g., groupLeader, group, or GET_PLAYERBOT_AI()) might dereference nulls under rare conditions. Adding early returns or explicit null checks would make the logic more robust.

Flow Complexity:
The nested if/continue structure makes it hard to follow the master selection flow. Refactoring into smaller helper functions (like FindBestRealPlayerCandidate() or FindBestBotCandidate()) could make it easier to reason about.

Master Selection Logic:
playerMaster and newMaster are both used as potential results but overlap in purpose. Consolidating them into a single bestMaster variable might reduce confusion.

Battleground Conditions:
The battleground-specific checks (especially AV) are quite specialized. It might be worth centralizing this into a dedicated helper for BG selection, or at least documenting why AV is treated differently.

Performance Considerations:
For large groups (like in AV), repeatedly calling functions like GET_PLAYERBOT_AI() and IsInSameRaidWith() for every member might add up. We could cache or reorder some checks for efficiency.

IsInSameRaidWith reliability:
Depending on the core version, this function can behave inconsistently for non-raid parties. Might be worth confirming or simplifying the check (e.g., comparing GetGroup() pointers directly).

Readability:
Adding some comments or clearer naming (like bestRealPlayer, bestBotFallback) would make intent more obvious for future maintainers.

None of these are urgent, but addressing them gradually could make this logic easier to extend and debug later on.

Zerathane commented:
I ran through some scenarios similar to ScoobyPwnsOnU and found the same results. The PR looks to resolve the issue of Rndbots following the player when leaving group/raid, Furthermore if a second real player is in the party the bots will switch to player to as their master if player one leaves.

I tested a few instances of Alterac Valley and found the bug seems to persist however. When the player leaves the battleground using the leave battleground button on the minimap, bots in the players party of the BG will still see the player as their master, despite being in AV and not being grouped with the player. 

The image below shows a bot identifying the player as a master despite being in AV. I was not grouped with the bot.
<img width="640" height="287" alt="botpartypr" src="https://github.com/user-attachments/assets/28ac7a91-7312-4135-adce-6e13c7399d3f" />

hermensbas commented:
Can some describe the exact issues we trying to solve. Just to ensure we all on same page and looking in the same direction. I see alot of issues described which are not linked to world packets leaving group and some might be? Bots leaving a party and still following you around for example is not, this was case before and after the world packets impl. 

So which exact issues are trying to solve?

Celandriel commented:
So the issue that prompted this was random bots not resetting properly on leaving group.

Alts and addclass are not considered because with no master they just get stuck. 

Resolving this issue identified a series of conflicting events related bots receiving packets, game state changes, and internal Ai updates.

A player leaving a group triggers the server sending a PARTY_OP_LEAVE message to be handled by the bot, while also destroying the group if that player is the leader. So what can happen is the internal update process could cause group updates at the updateaigroupmembership function or donextaction, or handling the world packet.

Then as I was editing I thought the scope of the Leave in leavegroupaction was too broad, because it was adding on another layer of verifications that could fail in a function that is used in multiple places. I simplified it to it handles random bots, and in the future we can also handle alts and addclass logging out. The conditional checking is moved to the main members of the class and depends on the message received whether the bot should follow it. One such condition was trying to find a new master before leaving the group, which led to a new handler FindNewMaster being created from existing code in DoNextAction. 

For the leaveordisbandgroup the only issue here was that sometimes by the time this function was processed the bots were no longer in the group, and the resets wouldn't happen because the first conditional was a reset check.
So I split it off into three sections leave group, reset master, general resetting.

Then for the final refactoring it didn't makes sense to me to have master and group checking occur twice in every Ai tick in both updateaiinternal and donextaction so the logic got merged in a unified place and I renamed the helper UpdateAiGroupAndMaster.

nl-saw commented:
Did some testing and the good news is that they now choose human leader if one present in the group, that is good. However, on Leave command, most of them now still stand still, and few just walk around a little. I'm pretty convinced this is due to all ```botAI->Reset();``` were removed. I think we need to keep it in one or more places.

Celandriel commented:
If you whisper 'who' to them, do they identify a master?

hermensbas commented:
shall be make it a draft meanwhile?

Celandriel commented:
Fixed the issue that nl-saw observed. 
The core problem was the leave group was prematurely resetting the master state before confirming that the bot is out of group. This caused the wrong set of strategies to be applied to the bot making them think they stayed in group on the strategy reset. The lack of master in this case limited the cleanup function UpdateAIGroupAndMaster to properly reset the master state once no longer in group. 

I confirmed random bots will swap masters to other real players as well.
